File: harness/lib/test-helpers/python/in_memory_repository.py
"""
InMemoryRepository genérico com isolamento por tenant_id.
Bug clássico de multi-tenant (esquecer o filtro) é impossível por design.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Generic, Optional, Protocol, TypeVar, runtime_checkable


@runtime_checkable
class TenantScoped(Protocol):
    id: str
    tenant_id: str


T = TypeVar("T", bound=TenantScoped)


@dataclass
class InMemoryRepository(Generic[T]):
    _store: dict[str, T] = field(default_factory=dict, init=False)

    async def find_by_id(self, id: str, tenant_id: str) -> Optional[T]:
        item = self._store.get(id)
        if item is None:
            return None
        return item if item.tenant_id == tenant_id else None

    async def save(self, item: T) -> None:
        self._store[item.id] = item

    async def delete(self, id: str, tenant_id: str) -> None:
        item = self._store.get(id)
        if item is not None and item.tenant_id == tenant_id:
            del self._store[id]

    async def find_all_by_tenant(self, tenant_id: str) -> list[T]:
        return [item for item in self._store.values() if item.tenant_id == tenant_id]

    async def find_all(self) -> list[T]:
        return list(self._store.values())

    async def count(self, tenant_id: Optional[str] = None) -> int:
        if tenant_id:
            return sum(1 for i in self._store.values() if i.tenant_id == tenant_id)
        return len(self._store)

    def reset(self) -> None:
        self._store.clear()


def create_in_memory_repository() -> InMemoryRepository:
    """Factory function para criar repositório in-memory."""
    return InMemoryRepository()

File: harness/lib/test-helpers/node/in-memory-repository.ts
export interface TenantScoped {
  id: string;
  tenantId: string;
}

export interface InMemoryRepository<T extends TenantScoped> {
  findById(id: string, tenantId: string): Promise<T | null>;
  save(item: T): Promise<void>;
  delete(id: string, tenantId: string): Promise<void>;
  findAllByTenant(tenantId: string): Promise<T[]>;
  findAll(): Promise<T[]>;
  count(tenantId?: string): Promise<number>;
  _reset(): void;
  _store(): Map<string, T>;
}

/**
 * Cria um repositório in-memory com isolamento por tenantId para uso em testes.
 * O filtro de tenantId é aplicado por padrão em todas as operações — bug clássico
 * de multi-tenant (esquecer o filtro) torna-se impossível.
 */
export function createInMemoryRepository<T extends TenantScoped>(): InMemoryRepository<T> {
  const store = new Map<string, T>();

  return {
    async findById(id: string, tenantId: string): Promise<T | null> {
      const item = store.get(id);
      if (!item) return null;
      return item.tenantId === tenantId ? item : null;
    },

    async save(item: T): Promise<void> {
      store.set(item.id, { ...item });
    },

    async delete(id: string, tenantId: string): Promise<void> {
      const item = store.get(id);
      if (item?.tenantId === tenantId) {
        store.delete(id);
      }
    },

    async findAllByTenant(tenantId: string): Promise<T[]> {
      return Array.from(store.values()).filter((item) => item.tenantId === tenantId);
    },

    async findAll(): Promise<T[]> {
      return Array.from(store.values());
    },

    async count(tenantId?: string): Promise<number> {
      if (tenantId) {
        return Array.from(store.values()).filter((i) => i.tenantId === tenantId).length;
      }
      return store.size;
    },

    _reset(): void {
      store.clear();
    },

    _store(): Map<string, T> {
      return store;
    },
  };
}
